applications/CompressiblePotentialFlowApplication/python_scripts/apply_far_field_process.py

```diff
@@ -15,65 +21,113 @@ def __init__(self, Model, settings ):
         default_parameters = KratosMultiphysics.Parameters( """
             {
                 "model_part_name":"PLEASE_CHOOSE_MODEL_PART_NAME",
-                "mesh_id": 0,
                 "inlet_phi": 1.0,
-                "velocity_infinity": [1.0,0.0,0]
-            }  """ );
-
-
-        settings.ValidateAndAssignDefaults(default_parameters);
-
-        self.model_part = Model[settings["model_part_name"].GetString()]
-        self.velocity_infinity = KratosMultiphysics.Vector(3)#array('d', [1.0, 2.0, 3.14])#np.array([0,0,0])#np.zeros(3)#vector(3)
-        self.velocity_infinity[0] = settings["velocity_infinity"][0].GetDouble()
-        self.velocity_infinity[1] = settings["velocity_infinity"][1].GetDouble()
-        self.velocity_infinity[2] = settings["velocity_infinity"][2].GetDouble()
-        #self.density_infinity = settings["density_infinity"].GetDouble() #TODO: must read this from the properties
-        self.inlet_phi = settings["inlet_phi"].GetDouble()
-        self.model_part.ProcessInfo.SetValue(CompressiblePotentialFlowApplication.VELOCITY_INFINITY,self.velocity_infinity)
-
-
+                "mach_infinity": 0.01,
+                "angle_of_attack": 0.0,
+                "density_infinity"  : 1.0,
+                "speed_of_sound": 340,
+                "gamma": 1.4,
+                "pressure_infinity": 101325,
+                "initialize_flow_field": true
+            }  """)
+
+        settings.ValidateAndAssignDefaults(default_parameters)
+
+        self.far_field_model_part = Model[settings["model_part_name"].GetString()]
+        self.fluid_model_part = self.far_field_model_part.GetRootModelPart()
+
+        self.inlet_phi_0 = settings["inlet_phi"].GetDouble()
+        self.density_inf = settings["density_infinity"].GetDouble()
+        self.mach_inf = settings["mach_infinity"].GetDouble()
+        self.gamma = settings["gamma"].GetDouble()
+        self.aoa = settings["angle_of_attack"].GetDouble()
+        self.pressure_inf = settings["pressure_infinity"].GetDouble()
+        self.a_inf = settings["speed_of_sound"].GetDouble()
+        self.initialize = settings["initialize_flow_field"].GetBool()
+
+        # Computing free stream velocity
+        self.u_inf = self.mach_inf * self.a_inf
+        self.velocity_inf = KratosMultiphysics.Vector(3)
+        self.velocity_inf[0] = round(self.u_inf*math.cos(self.aoa),8)
+        self.velocity_inf[1] = round(self.u_inf*math.sin(self.aoa),8)
+        self.velocity_inf[2] = 0.0
+
+        # For the model part
+        self.far_field_model_part.ProcessInfo.SetValue(CPFApp.VELOCITY_INFINITY, self.velocity_inf)
+
+        # For the conditions
+        self.fluid_model_part.GetProperties()[0].SetValue(CPFApp.DENSITY_INFINITY, self.density_inf)
+
+        # For the elements
+        self.fluid_model_part.GetProperties()[1].SetValue(CPFApp.VELOCITY_INFINITY, self.velocity_inf)
```

Comment:
You are assuming that your elements will always have a properties with id 1. This is a problem is this is not the case. You have the advantage of having a unique properties set, so you can take its id from the first element. Is not nice at all but safer than this.

Reply:
I also thought about this, but I really did not liked the idea of passing rCurrentProcessInfo so many times around. But I guess you are right and there is no other way. 

In terms of time and memory, what is better to do inside the element, this:

```c++
const array_1d<double, 3> vinfinity = GetProperties().GetValue(VELOCITY_INFINITY);
```

or this:
```c++
const array_1d<double, 3> vinfinity = rCurrentProcessInfo[VELOCITY_INFINITY];
```

Reply:
> I also thought about this, but I really did not liked the idea of passing rCurrentProcessInfo so many times around. But I guess you are right and there is no other way.

But this is how it is designed => check the interface of `Element` & `Condition` => there it is often the only argument being passed, which gives you quite some flexibility

Regarding the performance I think it it identical, afaik the both use the same object for this (DataValueContainer)

Reply:
Ok! Thanks for the info! I will change it to rCurrentProcessInfo then